File: go/pkg/stacks/stacks_1.go
package stacks

import (
	"fmt"
	"strconv"
	"strings"
)

// MinStack is LIFO data structure
// supporting O(1) push, pop, and min
type MinStack struct {
	Top      *MinFrame
	MinFrame *MinFrame
}

// ToString provides a string representation for a stack
func (s *MinStack) ToString() string {
	frames := []string{}
	top := s.Top

	for top != nil {
		frames = append(frames, top.ToString())
		top = top.Next
	}

	if len(frames) > 0 {
		ss := []string{"top"}
		ss = append(ss, frames...)
		ss = append(ss, "bottom")
		min := "None"

		if s.MinFrame != nil {
			min = strconv.Itoa(s.MinFrame.Data)
		} else {
			min = "nil"
		}

		return fmt.Sprintf("%s; Min: %s", strings.Join(ss, " -> "), min)
	}

	return "empty stack"
}

// Push O(1)
func (s *MinStack) Push(data int) {
	frame := &MinFrame{Data: data, Next: s.Top}

	if s.MinFrame != nil {
		if data < s.MinFrame.Data {
			frame.PreviousMinFrame = s.MinFrame
			s.MinFrame = frame
		}
	} else {
		s.MinFrame = frame
	}

	s.Top = frame
}

// Pop O(1)
func (s *MinStack) Pop() int {
	if s.Top == nil {
		panic("stack is empty")
	}

	if s.Top == s.MinFrame {
		s.MinFrame = s.Top.PreviousMinFrame
	}

	frame := s.Top
	s.Top = s.Top.Next

	return frame.Data
}

// Min O(1)
func (s *MinStack) Min() int {
	if s.Top == nil {
		panic("stack is empty")
	}

	return s.MinFrame.Data
}

// MinFrame represents a single stack frame for a MinStack
// using int for the data type out of convenience
type MinFrame struct {
	Data             int
	Next             *MinFrame
	PreviousMinFrame *MinFrame
}

// ToString provides a string representation for a stack frame
func (f *MinFrame) ToString() string {
	return fmt.Sprintf("[ %d ]", f.Data)
}

File: python/stacks/stacks_1.py
class MinStack:
    def __init__(self, top=None, min_frame=None):
        self.top = top
        self.min_frame = min_frame

    def __repr__(self):
        frames = []
        top = self.top

        while top:
            frames.append(str(top))
            top = top.next

        # [] is falsy in Python
        if frames:
            return f'{" -> ".join(["top"] + frames + ["bottom"])}; Min: {self.min_frame.data if self.min_frame else "None"}'

        return 'empty stack'

    def push(self, data):
        frame = MinFrame(data, self.top)

        # assumes that "data" is comparable to other "data" using standard operators
        if self.min_frame:
            if data < self.min_frame.data:
                frame.previous_min_frame = self.min_frame
                self.min_frame = frame
        else:
            self.min_frame = frame

        self.top = frame

    def pop(self):
        if not self.top:
            raise Exception('stack is empty')

        if self.top is self.min_frame:
            self.min_frame = self.top.previous_min_frame

        frame = self.top
        self.top = self.top.next

        return frame.data

    def min(self):
        if not self.top:
            raise Exception('stack is empty')

        return self.min_frame.data


class MinFrame:
    def __init__(self, data=None, next=None, previous_min_frame=None):
        self.data = data
        self.next = next
        self.previous_min_frame = previous_min_frame

    def __repr__(self):
        return f'[ {self.data} ]'
